core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -144,8 +164,24 @@ public long acquireExecutionMemory(long required, MemoryConsumer consumer) {
       // spilling, avoid to have too many spilled files.
       if (got < required) {
         // Call spill() on other consumers to release memory
+        // Sort the consumers according their memory usage. So we avoid spilling the same consumer
+        // which is just spilled in last few times and re-spilling on it will produce many small
+        // spill files.
+        List<MemoryConsumer> sortedList = new ArrayList<>(consumers.size());
         for (MemoryConsumer c: consumers) {
           if (c != consumer && c.getUsed() > 0 && c.getMode() == mode) {
+            sortedList.add(c);
+          }
+        }
+        Collections.sort(sortedList, new ConsumerComparator());
+        for (int listIndex = 0; listIndex < sortedList.size(); listIndex++) {
+          MemoryConsumer c = sortedList.get(listIndex);
+          // Try to only spill on the consumer which has the required size of memory.
+          // As the consumers are sorted in descending order, if the next consumer doesn't have
+          // the required memory, then we need to spill the current consumer at least.
+          boolean doSpill = (listIndex + 1) == sortedList.size() ||
+            sortedList.get(listIndex + 1).getUsed() < (required - got);
+          if (doSpill) {
```

Comment:
I am not sure how we keep iterating the consumers in existing impl. As we loop on all consumers, no matter we spill it or not or how many memory we released, we won't go back to it and previous consumers, right?